File: notification-hubs-sample-app-java/src/main/java/com/example/notification_hubs_sample_app_java/ui/main/NotificationListViewModel.java
package com.example.notification_hubs_sample_app_java.ui.main;

import android.app.Notification;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.notification_hubs_sample_app_java.cookbook.NotificationDisplayer;
import com.microsoft.windowsazure.messaging.notificationhubs.NotificationMessage;
import com.microsoft.windowsazure.messaging.notificationhubs.NotificationHub;
import com.microsoft.windowsazure.messaging.notificationhubs.NotificationListener;

import java.util.ArrayList;
import java.util.List;

public class NotificationListViewModel extends ViewModel {
    private final MutableLiveData<List<NotificationMessage>> mNotifications;
    private final NotificationListener mListener;

    public NotificationListViewModel() {
        mNotifications = new MutableLiveData<List<NotificationMessage>>();
        mListener = (context, message) -> {
            addNotification(message);
        };

//        NotificationHub.setListener(mListener);
        NotificationHub.setListener(new NotificationDisplayer());
    }

    public void addNotification(NotificationMessage notification) {
        List<NotificationMessage> toUpdate = mNotifications.getValue();
        if(toUpdate == null) {
            toUpdate = new ArrayList<NotificationMessage>();
        }
        toUpdate.add(0, notification);
        mNotifications.postValue(toUpdate);
    }

    public void clearNotifications() {
        List<NotificationMessage> toUpdate = mNotifications.getValue();
        if (toUpdate == null) {
            toUpdate = new ArrayList<NotificationMessage>();
        } else {
            toUpdate.clear();
        }
        mNotifications.postValue(toUpdate);
    }

    public LiveData<List<NotificationMessage>> getNotificationList() {
        return mNotifications;
    }
}

File: notification-hubs-test-app-kotlin/src/main/java/com/example/notification_hubs_test_app_kotlin/ui/main/NotificationListViewModel.kt
package com.example.notification_hubs_test_app_kotlin.ui.main

import android.content.Context
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.microsoft.windowsazure.messaging.notificationhubs.NotificationHub
import com.microsoft.windowsazure.messaging.notificationhubs.NotificationListener
import com.microsoft.windowsazure.messaging.notificationhubs.NotificationMessage

class NotificationListViewModel : ViewModel() {
    private val mNotifications: MutableLiveData<ArrayList<NotificationMessage>> = MutableLiveData()
    private val mListener: NotificationListener
    fun addNotification(notification: NotificationMessage?) {
        var toUpdate: ArrayList<NotificationMessage>? = mNotifications.value
        if (toUpdate == null) {
            toUpdate = ArrayList()
        }
        if (notification != null) {
            toUpdate.add(0, notification)
        }
        mNotifications.postValue(toUpdate)
    }

    fun clearNotifications() {
        var toUpdate: ArrayList<NotificationMessage>? = mNotifications.value
        if (toUpdate == null) {
            toUpdate = ArrayList()
        } else {
            toUpdate.clear()
        }
        mNotifications.postValue(toUpdate)
    }

    fun getNotificationList(): MutableLiveData<ArrayList<NotificationMessage>> {
        return mNotifications;
    }

    init {
        mListener = NotificationListener { context: Context?, message: NotificationMessage? -> addNotification(message) }
        NotificationHub.setListener(mListener)
    }
}